cores/esp32/IPAddress.cpp

```diff
@@ -117,9 +135,122 @@ bool IPAddress::fromString(const char *address)
         // Too few dots (there must be 3 dots)
         return false;
     }
-    _address.bytes[3] = acc;
+    (*this)[3] = acc;
+
+    setV4();
     return true;
 }
 
+IPAddress& IPAddress::operator=(const uint8_t *address) {
+    uint32_t value;
+    memcpy_P(&value, address, sizeof(value));
+    *this = value;
+    return *this;
+}
+
+IPAddress& IPAddress::operator=(uint32_t address) {
+    setV4();
+    v4() = address;
+    return *this;
+}
+
+bool IPAddress::operator==(const uint8_t* addr) const {
+    if (!isV4()) {
+        return false;
+    }
+
+    uint32_t value;
+    memcpy_P(&value, addr, sizeof(value));
+
+    return v4() == value;
+}
+
+size_t IPAddress::printTo(Print& p) const {
+    size_t n = 0;
+
+#if LWIP_IPV6
+    if (isV6()) {
+        int count0 = 0;
+        for (int i = 0; i < 8; i++) {
```

Comment:
Is this supposed to output the canonical format, i.e. with consecutive zeros of the longest run (not necessarily the first) combined?

Reply:
There is a ip6addr_ntoa() that will do the convert for us (and we then copy to the print buffer).

Reply:
There is also an `IPAddress:printTo()` function in https://github.com/arduino/ArduinoCore-API/blob/master/api/IPAddress.cpp which has the correct IPv6 IETF canonical format to compress the left-most longest run of two or more zero fields.

It does two passes... first start from the left to find a run of two or more zero fields, only replacing if a later run is longer (not the same). This gives a single location of the left-most longest run. The second pass then outputs, adding the compression at the detected location.

... but if we have access to ip6addr_ntoa() it might be better to just use that.

Reply:
I also suggest adding an automated test project. The ArduinoCore-API has automated tests for a whole bunch of corner cases of canonical to string, at https://github.com/arduino/ArduinoCore-API/blob/master/test/src/IPAddress/test_printTo6.cpp

e.g. where there is a run of 2 zeros followed by a later run of 3 zeros.

Reply:
This is a cut and paste from ESP8266. I have not experienced any issue in IPv6 formatting. Do you have a breaking example?

Reply:
A tricky example, where the last run of zeros is longest is: ip(0,0, 0,1, 0,0, 0,0, 0,2, 0,0, 0,0, 0,0). This should output: "0:1:0:0:2::"

Executing the current algorithm by hand gives ":1:0:0:2:0:0:0" (I think)

Some other tricky ones to check

ip(0,0, 0,1, 0,0, 0,0, 0,0, 0,2, 0,0, 0,0); => "0:1::2:0:0"
ip(0,0, 0,2, 0,3, 0,4, 0,5, 0,6, 0,7, 0,8); => "0:2:3:4:5:6:7:8"
ip(0,1, 0,2, 0,3, 0,4, 0,5, 0,6, 0,7, 0,0); => "1:2:3:4:5:6:7:0"
ip(0,0, 0,0, 0,3, 0,4, 0,5, 0,6, 0,7, 0,0); => "::3:4:5:6:7:0"
ip(0,0, 0,2, 0,3, 0,4, 0,5, 0,6, 0,0, 0,0); => "0:2:3:4:5:6::"
ip(0,1, 0,0, 0,0, 0,4, 0,5, 0,0, 0,0, 0,8); => "1::4:5:0:0:8" (both the same length, so compress the left-most)

If you have automated tests, just copy the cases from ArduinoCore